docs/design-proposals/http-public-repo.md

```diff
@@ -0,0 +1,123 @@
+
+The community requests an ability to upload charts in such a way that
+the index contains URLs with HTTP(S) schema instead of S3. This allows
+setting up a static file server that serves directly from the S3 bucket, so repository
+users can download charts directly using HTTP protocol.
+
+The key problem here is the index file. In the index, we can have either `s3://...`
+URLs, if using a repository as a private one, or `http(s)://...`, if we want a
+repository to be public. **helm** allows having multiple URLs for one chart in
+the index, but if we want a public repository - we don't even want to expose
+that we are using S3 as a backend.
+
+The index is static. We cannot upload charts setting `s3://...` URLs and
+then change them to corresponding `http(s)://...` on the fly. If index is
+served on HTTP(S), then **helm** downloads the index using HTTP(S) protocol
+and **helm-s3** is not involved here, so it cannot "inject" into the process
+and replace chart urls from `s3://...` to `http(s):/...`.
+
+## Proposal
+
+The proposed solution is to add metadata to the index file, so that **helm-s3** plugin
+can distinguish public and private repositories and make correct decisions.
+
+Metadata is managed through AWS S3 object metadata.
+
+The sections below describe the full use case from both point of views:
+the repository owner (developer) and the user.
+
+Note that examples below assume that the bucket `s3://awesome-bucket`  is
+already configured to serve files under `/charts` key on `https://charts.my-company.tld`.
+
+### Repository owner
+
+1. Initialize the repository.
+
+    ```
+    helm s3 init s3://awesome-bucket/charts --publish https://charts.my-company.tld
+    ```
+
+    1. **helm-s3** creates an empty `index.yaml` file and uploads it by path `s3://awesome-bucket/charts/index.yaml`.
+    It also adds a metadata to index s3 object with the key `helm-s3-public-repo` and value `https://charts.my-company.tld`.
+
+1. Add the repository locally to work with it.
+
+    ```
+    helm repo add my-charts s3://awesome-bucket/charts
+    ```
+
+    *note that developer (repository owner) adds the repo using s3 protocol*
+
+    1. **helm** downloads index from `s3://awesome-bucket/charts/index.yaml` and puts it into the local cache `/Users/me/.helm/repository/cache/my-charts-index.yaml`.
+
+1. Push chart to the repository.
+
+    ```
+    helm s3 push foo-0.5.2.tgz my-charts
+    ```
+
+    1. **helm-s3** uploads the file to the s3 bucket by path `s3://awesome-bucket/charts/foo-0.5.2.tgz`.
+
+    1. **helm-s3** updates the remote index, setting the url to public one from metadata `https://charts.my-company.tld/foo-0.5.2.tgz`
+    and not the `s3://...` url, because the index contains metadata informing that
+    the repo is public.
+
+    1. **helm-s3** also syncs the local cached index `/Users/me/.helm/repository/cache/my-charts-index.yaml` with the remote one.
+    Note that because of that, charts are fetch-able by public urls locally.
```

Comment:
How would authenticated repositories be handled? I'm using a private S3 bucket and exposing it to clients over https with basic authentication. If I use S3 plugin it doesn't store auth info for the publish URL. Perhaps local should keep S3 URLs?